City-of-Helsinki/berth-reservations
# VEN-1360 | Application changelog

## Description :sparkles:
* Add an `ApplicationChange` model that holds entries of changes to the values of applications
* Create an entry whenever an application is saved
* Create an entry whenever the `Update` mutation is called and the harbor/area choices change
* Small refactor to the logic of how `remove_choices` work, instead of passing the choice id, pass the priority to make it clearer

## Issues :bug:
### Closes :no_good_woman:
**[VEN-1360](https://helsinkisolutionoffice.atlassian.net/browse/VEN-1360):** Add changelog to application

## Testing :alembic:
### Automated tests :gear:️
```shell
$ pytest applications/tests/test_applications_mutations.py::test_update_winter_storage_application_by_owner
$ pytest applications/tests/test_applications_mutations.py::test_update_berth_application_by_owner
```

### Manual testing :construction_worker_man:
1. Change any value on an application, either through the API or the Django admin
2. Check that the change entry was created, either through the API (`berthApplication.changes`) or the Django Admin

## Screenshots :camera_flash:
**Django Admin:**
<img width="836" alt="image" src="https://user-images.githubusercontent.com/15201480/122995571-638dcc80-d36f-11eb-849c-4cec1baa6f78.png">

**GraphQL API:**
<img width="832" alt="image" src="https://user-images.githubusercontent.com/15201480/122995389-2aedf300-d36f-11eb-9344-34a5b6e3e024.png">
